### DISCUSSION TITLE
Contextual Active Filters on Product Grids UX Edition

### DISCUSSION THREAD

**Discussion by roykho** [woocommerce/woocommerce#44954] [WooCommerce, Blocks, and Site Editing]
In this discussion, I wanted to address a current issue with have with active filters (Filter by Stock, Filter by Price...etc) as it relates to product grids (Products (beta), Product Collection (beta)...etc).

For the rest of this discussion, I will be referring to only the `Filter by Stock` block and the `Product Collection` block for simplicity and use `filter, active filter(s)` interchangeably with `Filter by Stock` and `product grid block` interchangeably with `Product Collection` block.

### Current Problem

When we have an active filter such as the `Filter by Stock` block in a presence of a `Product Collection` block that utilizes the `Handpicked filters` setting, we encounter a problem. The problem is that the `Filter by Stock` block is using the query data from [collection data API](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/context/hooks/collections/use-collection-data.ts) without the context of the `Product Collection` block so it doesn't know to only query products that is selectively chosen in the `handpicked` settings. I have included screenshots of what I mean.

**Handpicked products setting:**
![file.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/41ffff1d-1f98-4554-ae98-a39c9cb1b15b)

**Frontend result**
![file.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/e4fcbfea-b481-4709-8bfe-bda2e7d793a4)

As you can see there are only 3 products that were handpicked but yet in the frontend, the filter is showing count results for all products in the store. In addition, the `Out of stock` filter shouldn't be showing up as there were no products that are out of stock in the handpicked products. Another words the filter has no context of the products actually being displayed.

### What Can We Do?

While researching this issue, I quickly realize UX will probably play a big part in the implementation. What does the user expect to happen in different situations?

Here are some UX questions I came up with:
- Do users expect adding one filter on a page to filter multiple product grid blocks?
- Do users expect filters to should only work on the main shop page?
- Do users expect each product grid block should have its own filters?

Here are some additional questions that may dictate how we approach this:
- How many users actually use product grid blocks outside of the main shop page?
- And to the above, if so, are they just adding selective products to show and not actually trying to display the whole product grid again (as it wouldn't make much sense as it is a repeat of their shop page)?

### Solution A

Since active filters in itself does not function without the presence of a product grid block, we can make active filters only insertable in the presence of a product grid block. This way the active filters have direct context from the parent product grid block to know how to formulate the query to hydrate only what we need.

| Pros | Cons |
| --- | --- |
| Active filters will have direct context to the parent product grid block | A little less flexibility with the placement of the active filter block as it now has to live within the parent product grid block |
| Cleaner and less confusing UX because it prevents user from adding an active filter block by itself and it not functioning without a product grid block | Since there can only be one filter per page (due to the limitation of only one set of the URL query string), it could cause confusion if user decides to add two product grid and two filters (probably not likely case though) |

### Solution B

Make the active filter only available to the main product grid block (ie. what would be displayed on the shop page).

| Pros | Cons |
| --- | --- |
| Cleaner and less confusing UX as filters only work with the main product grid block | User will no longer have ability to add an ad-hoc filter to a page with product grid block |

### In Closing

Personally I am partial to solution B as it is much easier to understand in terms of UX (imo) and implementation is also easier which means easier code maintenance.

I am sure there are other considerations that I have missed and would love any feedback that you may add. 

**Update**

Cross referencing a design discussion explicitly related to the filters block https://github.com/woocommerce/woocommerce-blocks/discussions/10691

**Comment by imanish003** [comment]
Hey @roykho,

Fantastic writeup! Building on the conversation, I'd like to introduce another potential solution.

**Solution C**: **Advanced Setting for Active Filters**

Instead of restricting the active filter with specific constraints, what if we embedded an advanced setting directly in the active filter blocks? This would empower users to explicitly select which product grid block(s) the active filter should correspond with.

- As an illustrative example: Perhaps we could incorporate a dropdown in the sidebar settings, allowing users to select their desired Product Grid. This concept could greatly benefit from design refinement, but it provides a starting point.

Interestingly, we already have distinct keys for Product Collection & Products beta block used for pagination. This ensures that:

- Even if we place 10 Product Collection blocks on the same page/post, the pagination for each block functions independently, thanks to the unique keys for different instances of the Product Collection block. Perhaps, this same key mechanism can be leveraged for our purpose here.

What are your thoughts on this?

**Reply by roykho** [reply]
Hi @imanish003 thanks for chiming in.

Yes I actually thought about that solution as well but quickly dismissed it as I haven't seen any stores personally that uses more than one set of product grids with multiple filters.

But here are other reasons I didn't list this:

- From a UX perspective, how would it look to select the product grid you want to target for that specific filter? You would probably have to grab the list of `client ids` on the page and have a drop down on the filter setting. But as a user, those `client ids` are just some strange alpha numeric string. How do they associate that with the product grid without switching to the code editor.
- How would the URL string look when they actually go to filter something. For example if they have two product grids and two stock filters. When they go to choose `In-stock` filter for the first set of product grid, the URL changes to for example `https://store.com/?filter_stock_status=instock`. That drives the query of all filters on the page and not any specific one. Of course you can do it with `?filter_stock_status_1=instock&filter_stock_status_2=outofstock`..etc but seems messy?
- Might be confusing for the customers to see multiple of the same named filters on the same page even though there are multiple product grids as well.

**Reply by imanish003** [reply]
Hi @roykho,

I'm glad to know that you thought about this solution as well. 

> • From a UX perspective, how would it look to select the product grid you want to target for that specific filter? You would probably have to grab the list of `client ids` on the page and have a drop down on the filter setting. But as a user, those `client ids` are just some strange alpha numeric string. How do they associate that with the product grid without switching to the code editor.
> 

I'm not entirely certain about this either. I view this primarily as a user experience challenge, and Shaun might be able to provide some insights. Here are some thoughts:

- Perhaps we could display names for each Product Collection block in the sidebar settings, which could then be shown in the dropdown.
- Maybe we can visually highlight the Product Collection block based on the selected value?

Nonetheless, I do agree that these ideas don't seem quite right, and they might not be practical or easy to understand.

> • How would the URL string look when they actually go to filter something. For example if they have two product grids and two stock filters. When they go to choose `In-stock` filter for the first set of product grid, the URL changes to for example `https://store.com/?filter_stock_status=instock`. That drives the query of all filters on the page and not any specific one. Of course you can do it with `?filter_stock_status_1=instock&filter_stock_status_2=outofstock`..etc but seems messy?
> 

I wouldn't describe this approach as messy. We're already employing a similar technique for pagination in the Product Collection, Products beta, and Query Loop blocks. Additionally, I believe we'll need to implement something similar for Shaun's suggestion in [Introduce a new Filters block woocommerce/woocommerce-blocks#10691](https://github.com/woocommerce/woocommerce-blocks/discussions/10691). Without incorporating state in the URL, we risk losing it if a user decides to refresh the page. 

---

I concur that this solution isn't ideal. It could potentially confuse users and prove challenging for us to implement. I'm quite intrigued by what Shaun shared in [Introduce a new Filters block woocommerce/woocommerce-blocks#10691](https://github.com/woocommerce/woocommerce-blocks/discussions/10691). It might be a good idea to abandon this solution in favor of Shaun's proposal.

**Comment by shaunandrews** [comment]
It seems to me that filters should be coupled with the collection they are affecting, and that most often a page will contain one set of filters that affects a single collection. Do we know of any real-world examples where having filters _outside_ of the collection block would be advantageous?

**Reply by roykho** [reply]
> and that most often a page will contain one set of filters that affects a single collection

That is how I imagine it as well and have personally seen in other various online shops.

> Do we know of any real-world examples where having filters outside of the collection block would be advantageous?

So the discussion meant the ability to add the filter block within the product grid block vs outside of it. Here is a screenshot that helps to explain it better:

| Outside | Inside |
| --- | --- |
| ![file.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/110de439-f57f-4458-8c29-349a996f0157) | ![file.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/768f07c8-1913-451b-abf4-14edce40b557) |

**Reply by shaunandrews** [reply]
Yup, understood. I'm asking what scenario's would benefit from having it "outside" rather than "inside?" My instincts tell me that it should always (and only) be available inside the Collection block.

**Reply by roykho** [reply]
Yes agreed. Originally I was going to say you have more flexibility in terms of positioning the filter wherever you want. However having just played with it, you can easily add columns and such inside the Product Collection block and still put the filter wherever you want. So I guess the answer to your question is "not much or if any".

**Reply by shaunandrews** [reply]
That seems to settle it then, for me at least; The filters should only be available within a Collection block. Here's a look at the current design I'm working on for these Inner Blocks:

<img width="2184" alt="CleanShot 2023-08-22 at 12 54 22@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/191598/85dc070a-ce23-4689-bd2e-c8700e7684ea">

**Reply by sunyatasattva** [reply]
Just came back from my AFK and trying to catch up will all the discussions. I apologize in advance if this was tackled already: I agree with scoping filter blocks within Product Collections in general, however, to play devil's advocate to @shaunandrews's question:

>  I'm asking what scenario's would benefit from having it "outside" rather than "inside?"

Imagine this scenario:

As a merchant, when the user navigates on a specific category, I want them to see the “Products on Sale” from that category at the top of the page, and then the entire catalog for that category.

Should the users be able to change only one set of filters (say, I only want to see blue stuff) and have it reflected in both the collections?

Again, playing devil's advocate here. I think this would introduce a bunch of extra UX complexity to have global and local filters (potentially conflicting with each other), implementation hurdles, and it could be confusing in exchange for the end-user to be able to cover a relatively edge case.

Nevertheless, I wanted to bring that up for discussion because I think there exist a few legitimate use-cases for this. So, if we don't support them, it'd be nice to be explicit about it.

**Reply by roykho** [reply]
> Should the users be able to change only one set of filters (say, I only want to see blue stuff) and have it reflected in both the collections?

I think it is rare to have two sets of collections on the same page however what is proposed here still works as you can have the Collection Filters block in each of the Products Collection. Our current pagination block works the same way. Also this is why having it "inside" the Products Collection is beneficial because now we have direct context of the query.

**Reply by sunyatasattva** [reply]
Yea sorry perhaps I should have worded my use-case more clearly: as you mentioned, this works by having two collection filters, one for each Products Collection block. What I meant to say, is that perhaps there is a case to be made for a global filter: if I'm interested in only “blue” things, I probably am interested in only “blue” things for both of the collections there, and I'd rather have to filter once (on the UX side) and also not have seemingly duplicate UI which might clutter the page.

Again, NBD, but I think we should just keep these things in mind.

**Reply by shaunandrews** [reply]
My assumption is that most people will have a "primary" collection, and perhaps additional collections. In the scenario you laid out, the "on sale" collection is a list of recommendations and would be unlikely to have it's own filters. The primary collection would have it's own set of filters, but they wouldn't apply to the "on sale" collection. We might consider hiding irrelevant collections if a filter is applied.

One example of this in the wild is on BestBuy.com. They show a list of featured products at the top, followed by their full catalog. Checking a filter in their sidebar then hides the list of featured products.

<img width="1523" alt="CleanShot 2023-08-28 at 10 35 04@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/191598/39c01934-5c14-439a-978e-f50c656bf7f3">

<img width="1509" alt="CleanShot 2023-08-28 at 10 35 26@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/191598/faed5958-33e7-4c55-9d42-63e737d0eed4">

**Reply by sunyatasattva** [reply]
Excellent! Thanks a lot @shaunandrews for replying with an actual real-world use-case example of this pattern. Seems sensible to me!
